include/elfloader/elfloader.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ELFLOADER_H
+#define ELFLOADER_H
+
+#include <elf.h>
+
+/* module exported symbol structure */
+typedef struct zmodule_symbol {
+	elf_addr addr;
+	const char *name;
+} zmodule_symbol_t;
+
+/* module structure */
+typedef struct zmodule {
+	elf_addr virt_start_addr;
+	elf_addr load_start_addr;
+	elf_word mem_sz;
+	zmodule_symbol_t *sym_list;
+	elf_word sym_cnt;
+} zmodule_t;
+
+/*
+ * Load a module into memory dynamically
+ */
+void *zmodule_load(const char *filename);
+
+/*
+ * Find arbitrary symbol's address according to its name.
+ * If module pointer is 0, then it will find symbols from zephyr kernel.
+ */
+void *zmodule_find_sym(zmodule_t *module, const char *sym_name);
```

Comment:
Can we get away from an explicit string in the API?  I mean, I know that right now this is a loader for ELF files in Zephyr, and ELF files have string symbol tables.  But Zephyr has always leaned very heavily on preprocessing for size optimization, and it would be sad to be stuck with this forever.

How about an opaque "`zmodule_sym_t`" type or something here?  Then you could have a constructor like `ZMODULE_SYM("my_funtion_name")` to build one that you could pass to this function.  Right now, that could be a noop that just evaluated to the.  But then later on we could come in and replace the module format with one that used pre-generated integer handles or something and nothing would break for applications.

Reply:
Oh, and for similar reasons, I'd call this header "zmodule.h" and not "elfloader.h".  We'll surely always be using ELF at the toolchain level, I'm not so sure it's the best fit for interpretation inside Zephyr itself.